composer/datasets/streaming/format.py

```diff
@@ -0,0 +1,286 @@
+
+        Returns:
+            cls: The loaded object.
+        """
+
+        dtype = np.int64
+        total_samples, total_bytes, num_shards, num_fields = read_array(fp, 4, dtype)
+        del total_bytes
+        samples_per_shard = read_array(fp, num_shards, dtype)
+        bytes_per_shard = read_array(fp, num_shards, dtype)
+        bytes_per_sample = read_array(fp, total_samples, dtype)
+        bytes_per_field = read_array(fp, num_fields, dtype)
+        fields = [fp.read(size).decode('utf-8') for size in bytes_per_field]
+        return cls(samples_per_shard, bytes_per_shard, bytes_per_sample, fields)
+
+    def dumps(self) -> bytes:
+        """Dump a StreamingDatasetIndex to raw bytes.
+
+        Returns:
+            bytes: The serialized form.
+        """
+        header = np.array([self.total_samples, self.total_bytes, self.num_shards, self.num_fields], np.int64)
+        bytes_per_field = np.array(list(map(len, self.fields)), np.int64)
+        arrays = header, self.samples_per_shard, self.bytes_per_shard, self.bytes_per_sample, bytes_per_field
+        arrays = np.concatenate(arrays, dtype=np.int64).tobytes()
+        fields = b''.join(map(lambda s: s.encode('utf-8'), self.fields))
+        return arrays + fields
+
+    def dump(self, fp: BufferedWriter) -> None:
+        """Dump a StreamingDatasetIndex to the file.
+
+        Args:
+            fp (file): The file to write.
+        """
+        data = self.dumps()
+        fp.write(data)
+
+    def _locate_samples(self) -> Tuple[NDArray[np.int64], NDArray[np.int64]]:
+        """Precompute the shard and byte offset within the shard of every sample.
+
+        Returns:
+            sample_shards (NDArray[np.int64]): Shard per sample.
+            sample_shard_offsets (NDArray[np.int64]): Intra-shard byte offset per sample.
+        """
+        shard_ends = self.bytes_per_shard.cumsum()
+        shard_begins = shard_ends - self.bytes_per_shard
+
+        sample_shard_begins = []
+        sample_shards = []
+        for shard, (num_samples, shard_begin) in enumerate(zip(self.samples_per_shard, shard_begins)):
+            sample_shard_begins += [shard_begin] * num_samples
+            sample_shards += [shard] * num_samples
+        sample_shard_begins = np.array(sample_shard_begins, np.int64)
+        sample_shards = np.array(sample_shards, np.int64)
+
+        sample_ends = self.bytes_per_sample.cumsum()
+        sample_begins = sample_ends - self.bytes_per_sample
+        sample_shard_offsets = sample_begins - sample_shard_begins
+        return sample_shards, sample_shard_offsets
+
+    def get_partition(self, world: World, batch_size: Optional[int] = None) -> Tuple[List[int], int, int]:
+        """Get the shards and sample range of a given partition of the dataset.
+
+        Args:
+            world (World): Context about workers, devices, and nodes.
+            batch_size (int): Hint the batch_size that will be used on each device's DataLoader.
```

Comment:
I'm confused by what is meant by hint here. My intuition is that providing a hint could make a function run faster, but it should return the same result even if I don't provide the hint. I don't think this is the case?

Perhaps could you rename the variable and provide additional description / an example of what this does?

Reply:
So providing `batch_size` will only change the behavior if:
 * you pass the object to a DataLoader
 * num_workers > 1
 * 
and I think @hanlint suggested using the same arg name as the DataLoader, so that users can do:

```
batch_size = 32
dataset = StreamingDataset(..., batch_size=batch_size)
loader = DataLoader(..., batch_size=batch_size)
```

The more descriptive variable name would be `device_batch_size_hint`, but IDK how you feel about doing that in the name or just elaborating in the description.

Reply:
Agree it makes sense to keep the variable the same. I guess could you elaborate in the docstring on why this is needed? An example showing the difference of providing it vs not providing it would be very helpful.

Reply:
Oh yeah there should be a really big explanation at the `StreamingDataset` docstring but I can copy it here as well.